test/versioned/when/segments.test.js

```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const test = require('node:test')
+const tspl = require('@matteo.collina/tspl')
+
+const { removeModules } = require('../../lib/cache-buster')
+const assertSegments = require('../../lib/custom-assertions/assert-segments')
+const helper = require('../../lib/agent_helper')
+
+// simulates a function that returns a promise and has a segment created for itself
+function doSomeWork({ tracer, Promise = global.Promise, segmentName, shouldReject } = {}) {
+  const segment = tracer.createSegment(segmentName)
+  return tracer.bindFunction(actualWork, segment)()
+
+  function actualWork() {
+    segment.touch()
+
+    return new Promise(function startSomeWork(resolve, reject) {
+      if (shouldReject) {
+        process.nextTick(function () {
+          reject('some reason')
+        })
+      } else {
+        process.nextTick(function () {
+          resolve(123)
+        })
+      }
+    })
+  }
+}
+
+test('segments enabled', async (t) => {
+  test.beforeEach((ctx) => {
+    ctx.nr = {}
+    ctx.nr.agent = helper.instrumentMockedAgent({
+      feature_flag: { promise_segments: true }
+    })
+    ctx.nr.tracer = ctx.nr.agent.tracer
+    ctx.nr.when = require('when')
+  })
+
+  test.afterEach((ctx) => {
+    helper.unloadAgent(ctx.nr.agent)
+    removeModules(['when'])
+  })
+
+  await t.test('child segment is created inside then handler', async (t) => {
+    const plan = tspl(t, { plan: 10 })
+    const { agent, tracer, when } = t.nr
+
+    agent.once('transactionFinished', (tx) => {
+      // doSomeWork => Promise startSomeWork => Promise#then <anonymous> => someChildSegment
+      plan.equal(tx.trace.root.children.length, 1)
+
+      assertSegments(
+        tx.trace.root,
+        [
+          'doSomeWork',
+          ['Promise startSomeWork', ['Promise#then <anonymous>', ['someChildSegment']]]
+        ],
+        {},
+        { assert: plan }
+      )
+    })
+
+    helper.runInTransaction(agent, function transactionWrapper(transaction) {
+      doSomeWork({ tracer, Promise: when.Promise, segmentName: 'doSomeWork' }).then(function () {
+        const childSegment = tracer.createSegment('someChildSegment')
+        // touch the segment, so that it is not truncated
+        childSegment.touch()
+        tracer.bindFunction(function () {}, childSegment)
+        process.nextTick(transaction.end.bind(transaction))
+      })
+    })
+
+    await plan.completed
+  })
+
+  // Note: this test was originally a `promise_segments: false` test listed
+  // under the segments enabled suite.
+  await t.test('then handler that returns a new promise', async (t) => {
```

Comment:
This test was originally https://github.com/newrelic/node-newrelic/blob/dffdac7f450a23c6e23a173c91bda583d40e573d/test/versioned/when/legacy-promise-segments.js#L67-L78

I don't think it should have been under the "promise_segments: true" group to begin with.